### PR TITLE
Upload OVF by URL

### PR DESCRIPTION
**Pull request by vbauzysvmware**
Ref: https://github.com/vmware/go-vcloud-director/issues/321

Add function to upload OVF by link. Only OVF files are supported.
Added support for updating vApp template `vAppTemplate.UpdateAsync` and `vAppTemplate.Update`

### CODE REVIEW

File: govcd/catalog.go [hunk 1]
```diff
@@ -250,6 +250,54 @@ func (cat *Catalog) UploadOvf(ovaFileName, itemName, description string, uploadP
 	return *uploadTask, nil
 }
 
+// UploadOvfByLink uploads an ovf file to a catalog from remote URL.
```
Comment (lvirbalas) [inline govcd/catalog.go:253]:
```suggestion
// UploadOvfByLink uploads an OVF file to a catalog from remote URL.
```

File: govcd/catalog.go [hunk 2]
```diff
@@ -250,6 +250,54 @@ func (cat *Catalog) UploadOvf(ovaFileName, itemName, description string, uploadP
...
+// Returns errors if any occur during upload from vCD or upload process. On upload fail client may need to
+// remove vCD catalog item which is in failed state.
```
Comment (lvirbalas) [inline govcd/catalog.go:255]: There are two nits in these two lines and you know them :)

File: govcd/catalog.go [hunk 3]
```diff
@@ -409,6 +454,20 @@ func queryVappTemplate(client *Client, vappTemplateUrl *url.URL, newItemName str
 	return vappTemplateParsed, nil
 }
 
+func fetchVappTemplate(client *Client, vappTemplateUrl *url.URL) (*types.VAppTemplate, error) {
+	util.Logger.Printf("[TRACE] Querying vapp template: %s\n", vappTemplateUrl)
```
Comment (lvirbalas) [inline govcd/catalog.go:458]:
```suggestion
	util.Logger.Printf("[TRACE] Querying vApp template: %s\n", vappTemplateUrl)
```

**Review by lvirbalas — CHANGES_REQUESTED** [review]
Few nits and a question: how much test time do the new tests add?

File: govcd/catalog.go (hunk 1 above)
Comment (vbauzysvmware) [reply govcd/catalog.go:253]: changed

File: govcd/catalog.go (hunk 2 above)
Comment (vbauzysvmware) [reply govcd/catalog.go:255]: updated VCD

File: govcd/catalog.go (hunk 3 above)
Comment (vbauzysvmware) [reply govcd/catalog.go:458]: fixed

**Comment by lvirbalas** [comment]
How much time do the new tests take?

**Comment by vbauzysvmware** [comment]
> How much time do the new tests take?

haven't checked, but should be similar as the file we use is the same and super small.

**Comment by lvirbalas** [comment]
> > How much time do the new tests take?
> 
> haven't checked, but should be similar as the file we use is the same and super small.

Then at least how much the total catalog_test.go time takes now?

**Comment by vbauzysvmware** [comment]
> > > How much time do the new tests take?
> > 
> > 
> > haven't checked, but should be similar as the file we use is the same and super small.
> 
> Then at least how much the total catalog_test.go time takes now?

New test run from my local env.

START: vapptemplate_test.go:52: TestVCD.Test_UpdateVAppTemplate
Running: TestVCD.Test_UpdateVAppTemplate
PASS: vapptemplate_test.go:52: TestVCD.Test_UpdateVAppTemplate	44.348s

Do you need times of catalog_test.go test functions, catalog tag(more tests than in catalog_test.go), or specific test methods times?

File: govcd/catalog_test.go [hunk 4]
```diff
@@ -788,3 +788,38 @@ func cleanupCatalogOrgVdc(check *C, sharedCatalog Catalog, vdc *Vdc, vcd *TestVC
 	err = adminOrg.Delete(true, true)
 	check.Assert(err, IsNil)
 }
+
+// Tests System function UploadOvf by creating catalog and
```
Comment (lvirbalas) [inline govcd/catalog_test.go:792]: I think this description is incorrect, because you're testing UploadOvfByLink here.

**Comment by lvirbalas** [comment]
> New test run from my local env.
> 
> START: vapptemplate_test.go:52: TestVCD.Test_UpdateVAppTemplate Running: TestVCD.Test_UpdateVAppTemplate PASS: vapptemplate_test.go:52: TestVCD.Test_UpdateVAppTemplate 44.348s

Thanks!

> Do you need times of catalog_test.go test functions, catalog tag(more tests than in catalog_test.go), or specific test methods times?

I'm actually most curious how long does the upload from URL test takes: `Test_UploadOvfByLink_progress_works`

File: govcd/catalog_test.go (hunk 4 above)
Comment (vbauzysvmware) [reply govcd/catalog_test.go:792]: updated

**Comment by vbauzysvmware** [comment]
> > New test run from my local env.
> > START: vapptemplate_test.go:52: TestVCD.Test_UpdateVAppTemplate Running: TestVCD.Test_UpdateVAppTemplate PASS: vapptemplate_test.go:52: TestVCD.Test_UpdateVAppTemplate 44.348s
> 
> Thanks!
> 
> > Do you need times of catalog_test.go test functions, catalog tag(more tests than in catalog_test.go), or specific test methods times?
> 
> I'm actually most curious how long does the upload from URL test takes: `Test_UploadOvfByLink_progress_works`

START: catalog_test.go:794: TestVCD.Test_UploadOvfByLink_progress_works
Running: TestVCD.Test_UploadOvfByLink_progress_works
PASS: catalog_test.go:794: TestVCD.Test_UploadOvfByLink_progress_works	42.877s

**Review by lvirbalas — APPROVED** [review]
Thanks for additional answers!

File: govcd/catalog.go [hunk 5]
```diff
@@ -520,6 +579,38 @@ func createItemForUpload(client *Client, createHREF *url.URL, catalogItemName st
 	return ovfUploadUrl, nil
 }
 
+// Initiates creation of item in catalog and returns vappTeamplate Url for created item.
+func createItemWithLink(client *Client, createHREF *url.URL, catalogItemName, itemDescription, vappTemplateRemoteUrl string) (*url.URL, error) {
+	util.Logger.Printf("[TRACE] createItemWithLink: %s, item name: %s, description: %s, vappTemplateRemoteUrl: %s \n", createHREF, catalogItemName, itemDescription, vappTemplateRemoteUrl)
+	reqBody := bytes.NewBufferString(
```
Comment (Didainius) [inline govcd/catalog.go:585]: It is very hard to read this template:
* Is it not possible to use a struct for that?
* Or at least a single template string in backticks `` and `Sprintf` to be something like (not ideal, but a bit more readable):

```go
	reqTemplate := `<UploadVAppTemplateParams xmlns="%s" name="%s" sourceHref="%s">
<Description>%s</Description>
</UploadVAppTemplateParams>
`
	body := fmt.Sprintf(reqTemplate, types.XMLNamespaceVCloud, catalogItemName, vappTemplateRemoteUrl, itemDescription)
	reqBody := bytes.NewBufferString(body)
```

File: types/v56/types.go [hunk 6]
```diff
@@ -1445,6 +1445,23 @@ type VAppTemplate struct {
 	// Section               Section              `xml:"Section,omitempty"`
 }
 
+// VAppTemplateForUpdate represents a vApp template.
+// Type: VAppTemplateType
+// Namespace: http://www.vmware.com/vcloud/v1.5
+// Description: Represents a vApp template.
+type VAppTemplateForUpdate struct {
```
[the remark this answers is not in the document]
Comment (vbauzysvmware) [reply types/v56/types.go:1452]: added

File: govcd/catalog.go (hunk 5 above)
Comment (vbauzysvmware) [reply govcd/catalog.go:585]: Good point, made similar to your idea.

**Review by Didainius — APPROVED** [review]

**Pull request merged**